### ISSUE TITLE
Lua filters should traverse the AST in linear order.

### ISSUE THREAD

**Issue by bpj** [jgm/pandoc#5667]
I'm creating this issue as per discussion in #1425, since nobody has done so yet, so that it isn't forgotten. (I guess @tarleb would like to be notified.)

The main use case (perhaps obvious) is probably filters which need to keep a count of visited elements and also use that count in the elements they add to the AST, e.g. id attributes and link targets, or which need to rely on elements being visited in the order in which they appear in the document for some other reason.

The filters for implementing endnotes vs. footnotes/per section notes and multiple apparatuses as per [manyfoot](http://texdoc.net/pkg/manyfoot) discussed in #1425 present yet another problem, beside needing to output note texts in the right order when generating HTML and other formats where note texts need to be put in an injected table element, even when producing LaTeX: since Note elements are inline elements containing block elements, you need to collect the note texts in each block in an array, replace the Note element with a RawInline element containing an `\endnotemark` or `\footnotemark<suffix>` command and then place the note texts corresponding to the note marks immediately after the block, wrapped in the appropriate RawBlock elements as well as a div for the benefit of subsequent filters. I have used this technique successfully in JSON filters written with [Pandoc::Walker](https://metacpan.org/pod/Pandoc::Walker) and even though `\footnotetext` and similar commands should ideally be placed within the paragraph it has hitherto worked to place them just after the paragraph. This however requires that you use "manually"/explicitly inserted footnote numbers, and this pretty much requires implementing a counter in the filter, which in turn requires linear traversal order.  I have tried to implement this in Lua by keeping a counter in Lua which is used to generate a TeX command which has the Lua counter number (converted to letters using the algorithm of [Number::Latin](https://metacpan.org/pod/Number::Latin)) in its name and expanding to the right note number obtained from a LaTeX counter, but it is ugly and far more complicated than it should be, and it also is confusing if the two counter sequences are out of synch. This involves a "main" filter which visits all blocks and a "local" filter closure which defines the above-mentioned number macros, collects footnote texts, and inserts footnote marks before the main filter function returns the original block and the footnote texts. I also don't know yet the effect of possibly getting the footnote texts in the wrong sequence. I won't be able to look into that until next week at the earliest.

**Comment by jgm** [comment]
We could fix this either by implementing custom walk logic in lua (rather than deferring to pandoc's walk), or by changing walk itself. The latter would also affect Haskell filters.

**Comment by tarleb** [comment]
My current approach is to build a separate traversal system which builds
on `walk` where possible. I'm not sure yet if this is a good direction.

**Comment by jgm** [comment]
It would be worth exploring how hard it would be to just change `walk`; that might be easier and more reliable in the end.

**Comment by bpj** [comment]
@jgm, @tarleb Thank you for looking into this so quickly! So if I
understand correctly there are two concerns: if the Haskell walk function
is changed Haskell filters will be affected and (I guess) if it is
implemented in Lua speed is affected, right?

Den ons 31 juli 2019 02:01John MacFarlane <notifications@github.com> skrev:

**Comment by tarleb** [comment]
Performance might not be affected much with a pure Lua solution. In fact, it might even be faster for a lot of cases: Lua is quite fast for a scripting language, while passing data from Lua to Haskell and back is slow.

Downsides of a pure Lua solution are: no type-safety, not re-usable outside of Lua filters, and it furthers the separation of Haskell and Lua code.

Upsides: potentially quicker to write, malleable and hackable.

**Comment by mb21** [comment]
> Upsides: potentially quicker to write, malleable and hackable.

🙈🤣

**Comment by tarleb** [comment]
**from within Lua** 🙃

**Comment by tarleb** [comment]
The following is currently my preferred solution:

``` haskell
-- | Helper type which allows us to traverse trees in order.
newtype SingletonsList a = SingletonsList { singletonsList :: [a] }
  deriving (Functor, Foldable, Monoid, Semigroup, Traversable)

instance {-# OVERLAPPING #-} Walkable (SingletonsList Inline) [Inline] where
  walkM f =
    let f' = fmap singletonsList . f . SingletonsList . (:[]) <=< walkM f
    in fmap mconcat . mapM f'
```

This gives us the desired properties without too many code changes. The major downside is that we'd have to either make this part of module `Text.Pandoc.Walk`, or export `walkPandocM`, `walkInlineM`, etc from that module. The latter is my current preference.

My idea of offering a node-by-node hierarchical tree traversal, which could be useful for now-and-then, should probably have it's own separate ticket.

**Comment by jgm** [comment]
I'd be okay with exporting these things from T.P.Walk.  I think that with a change that just adds exports, we could just increment the third version number and conform with PVP?

**Comment by tarleb** [comment]
That was my interpretation as well. I opened a PR at pandoc-types and will push the fixes here once that is merged.
